tensorboard/backend/client_feature_flags.py

```diff
@@ -39,25 +44,66 @@ def __init__(self, application):
         self._application = application
 
     def __call__(self, environ, start_response):
-        possible_feature_flags = environ.get("HTTP_X_TENSORBOARD_FEATURE_FLAGS")
-        if not possible_feature_flags:
+        header_feature_flags = self._parse_potential_header_param_flags(
+            environ.get("HTTP_X_TENSORBOARD_FEATURE_FLAGS")
+        )
+        query_string_feature_flags = self._parse_potential_query_param_flags(
+            environ.get("QUERY_STRING")
+        )
+
+        if not header_feature_flags and not query_string_feature_flags:
             return self._application(environ, start_response)
 
+        # header flags take precedence
+        for flag, value in header_feature_flags.items():
+            query_string_feature_flags[flag] = value
+
+        ctx = context.from_environ(environ).replace(
+            client_feature_flags=query_string_feature_flags
+        )
+        context.set_in_environ(environ, ctx)
+
+        return self._application(environ, start_response)
+
+    def _parse_potential_header_param_flags(self, header_string):
+        if not header_string:
+            return {}
+
         try:
-            client_feature_flags = json.loads(possible_feature_flags)
+            header_feature_flags = json.loads(header_string)
         except json.JSONDecodeError:
             raise errors.InvalidArgumentError(
                 "X-TensorBoard-Feature-Flags cannot be JSON decoded."
             )
 
-        if not isinstance(client_feature_flags, dict):
+        if not isinstance(header_feature_flags, dict):
             raise errors.InvalidArgumentError(
                 "X-TensorBoard-Feature-Flags cannot be decoded to a dict."
             )
 
-        ctx = context.from_environ(environ).replace(
-            client_feature_flags=client_feature_flags
+        return header_feature_flags
+
+    def _parse_potential_query_param_flags(self, query_string):
+        if not query_string:
+            return {}
+
+        try:
+            query_string_json = urllib.parse.parse_qs(query_string)
+        except ValueError:
+            return {}
+
+        # parse_qs returns the dictionary values as lists for each name.
+        potential_feature_flags = query_string_json.get(
+            "tensorBoardFeatureFlags", []
         )
-        context.set_in_environ(environ, ctx)
+        if not potential_feature_flags:
+            return {}
+        try:
+            client_feature_flags = json.loads(potential_feature_flags[0])
+        except json.JSONDecodeError:
+            return {}
 
-        return self._application(environ, start_response)
+        if not isinstance(client_feature_flags, dict):
+            return {}
```

Comment:
Should we also consider making this an InvalidArgumentError? Also similar to the header-parsing logic.

Reply:
Yep, here as well. Updated.